### PR TITLE
Small improvement in Sunken Temple

### PR DESCRIPTION
**Pull request by cala**
Shade of Eranikus is unattackable until Jammal'An the Prophet is defeated

### CODE REVIEW

**Comment by cala** [comment]
Suggested DB update:
```SQL
-- Updated NPC 5709 (Shade of Eranikus) in Sunken Temple: it is now unattackable
-- until Jammal'An is defeated
UPDATE `creature_template` SET `UnitFlags` = `UnitFlags` | 2 WHERE `Entry` = 5709;
```

See https://github.com/classicdb/database/issues/710 for more details.

**Comment by xfurry** [comment]
I'm not so sure about this....
First: as I can read in the classicdb issue, the dragon helpers can be handled by creature linking (and it's already done in UDB).
Second: the flag needs to be confirmed by sniff. This is very important because unit flags are one of the most important elements in the creature_template table. So any change in this field will have an impact on the behavior.
So unless this is confirmed by sniff I can't make this change.

I'm saying this, because from my experience, this unit flag 2 will always be accompanied by a neutral / friendly faction or unit flag 512. Otherwise it won't make sense, because the boss will attack the party if they get in range, but the players won't be able to attack the boss.

So, let's do a better research on this and check if there are other unit flags / factions / doors involved here.

**Comment by cala** [comment]
The dragon helpers are also handled in Classic DB, there were factions errors for some of them that are now fixed.

The only issue about Shade of Eranikus is that it is currently attackable before boss Jammal'An is defeated. It should not be.

At the moment, Shade of Eranikus has UnitFlags = UNIT_FLAG_UNK_6 (64) and ExtraFlags = CREATURE_FLAG_EXTRA_CIVILIAN (2). So players can come in contact with the boss, dance on its head... without it aggroing. This is by designed as this NPC is supposed to be sleeping, so it is unable to "notice" players. But at the moment, players can attack it at anytime while they should not, hence the use of UNIT_FLAG_NON_ATTACKABLE to prevent players from attacking it before due time.

But I understand your point and I'll try to get the right flag value. I think @evil-at-wow can be of some help here as he stated to have some sniffs about this NPC (id 5709).

**Comment by xfurry** [comment]
> But I understand your point and I'll try to get the right flag value. I think @evil-at-wow can be of some help here as he stated to have some sniffs about this NPC (id 5709).

Please.
I don't like the idea to change the things which are quite sensible without a proper documentation.

**Comment by evil-at-wow** [comment]
@cala Good that you explicitly remind me, because this wasn't on my list any more! I must have misread your reply about the force field and "are already in Classic DB", so I didn't plan any action...

But, good news, I do have some data for this and the oldest data I have is from early WoTLK (version 3.0.3, build 9183). I tracked Shade of Eranikus' guid through the data in an attempt to reconstruct the entire story, and what follows is basically what happens to him in the entire run (but limited to SMSG_COMPRESSED_UPDATE_OBJECT, SMSG_UPDATE_OBJECT and SMSG_UPDATE_OBJECT packets). Just for the sake of completeness, I included pretty much everything for the actual fight too.

First, the server notifies the client about the spawn:
```
Packed GUID = 0xf13000164d000171
Update flags = 0x78
Position (x y z) = (-659.994 -34.0331 -90.7518)
Orientation = 1.5708

Update field   0 = 0x7101004d
Update field   1 = 0x160030f1
Update field   2 = 0x09000000
Update field   3 = 0x4d160000  OBJECT_FIELD_ENTRY     (0x0000164d = 5709)
Update field   4 = 0x0000803f  OBJECT_FIELD_SCALE_X   (0x3f800000 = 1.0)
Update field  22 = 0x00010201  UNIT_FIELD_BYTES_0
Update field  23 = 0x7b700000  UNIT_FIELD_HEALTH
Update field  26 = 0x64000000  UNIT_FIELD_POWER3
Update field  27 = 0x64000000  UNIT_FIELD_POWER4
Update field  28 = 0x40420f00  UNIT_FIELD_POWER5
Update field  31 = 0x7b700000  UNIT_FIELD_MAXHEALTH
Update field  34 = 0x64000000  UNIT_FIELD_MAXPOWER3
Update field  35 = 0x64000000  UNIT_FIELD_MAXPOWER4
Update field  36 = 0x40420f00  UNIT_FIELD_MAXPOWER5
Update field  53 = 0x32000000  UNIT_FIELD_LEVEL
Update field  54 = 0x32000000  UNIT_FIELD_FACTIONTEMPLATE  (50)
Update field  58 = 0x40010000  UNIT_FIELD_FLAGS
Update field  59 = 0x00080000  UNIT_FIELD_FLAGS_2
Update field  61 = 0xd0070000  UNIT_FIELD_BASEATTACKTIME1
Update field  62 = 0xd0070000  UNIT_FIELD_BASEATTACKTIME2
Update field  64 = 0xcdcc4c3f  UNIT_FIELD_BOUNDINGRADIUS
Update field  65 = 0x00000041  UNIT_FIELD_COMBATREACH
Update field  66 = 0x7e1e0000  UNIT_FIELD_DISPLAYID
Update field  67 = 0x7e1e0000  UNIT_FIELD_NATIVEDISPLAYID
Update field  73 = 0x03000000  UNIT_FIELD_BYTES_1
Update field  80 = 0x0000803f  UNIT_MOD_CAST_SPEED
Update field 122 = 0x01000000  UNIT_FIELD_BYTES_2
Update field 146 = 0x0000803f  UNIT_FIELD_HOVERHEIGHT
```
As we can see, the faction template (50) seems to suggest "Dragonflight, Green". The UnitFlags are 0x0140 = UNIT_FLAG_UNK_6 + UNIT_FLAG_OOC_NOT_ATTACKABLE, and 
UnitFlags2 is 0x0800 = UNIT_FLAG2_REGENERATE_POWER.

**So you were pretty close, but it seems it's not UNIT_FLAG_NON_ATTACKABLE but UNIT_FLAG_OOC_NOT_ATTACKABLE instead (at least in WoTLK).**

Quite some time later, the server notifies the client again about the spawn, with the exact same data and fields as above, except for one single update field:
```
Update field 58 = 0x40000000  UNIT_FIELD_FLAGS
```
As far as I can see, there were *no* other updates in between. My guess is simply that I ran too far from Shade of Eranikus at one point when killing the troll bosses, then killed Jammal'an and headed back in the direction of Shade of Eranikus, triggering the server to resend me the needed data, but this time with the "not attackable" flag cleared (because Jammal'an was killed).

A bit later, new updates come in:
```
Packed GUID = 0xf13000164d000171
Update field 18 = 0xb6843a02  UNIT_FIELD_TARGET_LOW
Update field 19 = 0x00000000  UNIT_FIELD_TARGET_HIGH
Update field 58 = 0x40080000  UNIT_FIELD_FLAGS
Update field 73 = 0x00000000  UNIT_FIELD_BYTES_1
```
Basically, Shade of Eranikus was targetting someone (our tank I bet), and the unit flag
UNIT_FLAG_PET_IN_COMBAT was added (this could be caused by the "let's call all other dragons on the floor" behavior).

Immediately followed by:
```
Packed GUID = 0xf13000164d000171
Update field 58 = 0x40080800  UNIT_FIELD_FLAGS
```
The unit flag UNIT_FLAG_IN_COMBAT was added.
Then the client gets another update.
```
Packed GUID = 0xf13000164d000171
Update field 23 = 0x1b700000  UNIT_FIELD_HEALTH
```
The health of the boss has been updated.
```
Packed GUID = 0xf13000164d000171
Update field 78 = 0x0c000000  UNIT_DYNAMIC_FLAGS
Update field 82 = 0x00000000  UNIT_NPC_FLAGS
```
Aha! UnitDynFlags = UNIT_DYNFLAG_TAPPED + UNIT_DYNFLAG_TAPPED_BY_PLAYER! Why they are sending the NPC flags explicitly, I don't know.

Then, as you can expect from a boss fight, lots and lots of updates for health, once in a while a target update, and even one update of field 60 (UNIT_FIELD_AURASTATE). To end with this series of updates:
```
Packed GUID = 0xf13000164d000171
Update field 23 = 0x37000000  UNIT_FIELD_HEALTH
```
A health update.
```
Packed GUID = 0xf13000164d000171
Update field 18 = 0x00000000  UNIT_FIELD_TARGET_LOW
Update field 19 = 0x00000000  UNIT_FIELD_TARGET_HIGH
Update field 23 = 0x00000000  UNIT_FIELD_HEALTH
Update field 26 = 0x00000000  UNIT_FIELD_POWER3
Update field 27 = 0x00000000  UNIT_FIELD_POWER4
Update field 58 = 0x40080000  UNIT_FIELD_FLAGS
```
Clear the target, no health/mana/... remaining. He's dead! Also remove UNIT_FLAG_IN_COMBAT.
```
Packed GUID = 0xf13000164d000171
Update field 60 = 0x00000000  UNIT_FIELD_AURASTATE
Update field 78 = 0x0d000000  UNIT_DYNAMIC_FLAGS
Update field 82 = 0x00000000  UNIT_NPC_FLAGS
```
Add UNIT_DYNFLAG_LOOTABLE to UnitDynFlags basically.
```
Packed GUID = 0xf13000164d000171
Update field 58 = 0x40000000  UNIT_FIELD_FLAGS
```
Remove unit flag UNIT_FLAG_PET_IN_COMBAT.
```
Packed GUID = 0xf13000164d000171
Update field 78 = 0x0d000000  UNIT_DYNAMIC_FLAGS
Update field 82 = 0x00000000  UNIT_NPC_FLAGS
```
And end with repeating what's already been done...

**Comment by cala** [comment]
Wow at evil !! That is a pretty complete answer. Awesome!

Thank you for taking the time to do this. Though your data are from WotLK, I'm really confident they are also relevant to Classic and TBC. The only change made to Eranikus encounter between Classic and WotLK was about the helping dragons: only the four drakes in WotLK came to help while all the green drakes came back in Classic (maybe TBC).

So basically, in my current patch, I only need to change the flag removed (and change the DB patch accordingly).

Thanks. :+1:

**Comment by xfurry** [comment]
Well, in case of flag ```UNIT_FLAG_OOC_NOT_ATTACKABLE``` things are not that simple.
This flag will reset on evade. So in case of a wipe, the flag will be reset. Also, as I said, this won't prevent the boss from attacking the players on LoS. And this flag will be removed automatically in combat, so adding & removing it won't help in any way to our final goal.

In order to really fix this using this new flag, we need something like this: https://github.com/scriptdev2/scriptdev2/blob/master/scripts/northrend/crusaders_coliseum/trial_of_the_crusader/boss_faction_champions.cpp#L60

**Comment by cala** [comment]
>  Also, as I said, this won't prevent the boss from attacking the players on LoS.

NPC has ExtraFlags = CREATURE_FLAG_EXTRA_CIVILIAN in UDB, YTDB, TBC-DB and Classic DB, so it doesn't attack players on LoS and this is how this boss should behave. So unless you think we should do this otherwise, things looks OK here.

> in case of flag UNIT_FLAG_OOC_NOT_ATTACKABLE things are not that simple. This flag will reset on evade.

I was not aware of this and in this case, yes you are right: we must ensure that flag is kept removed even after a wipe.

**Comment by cala** [comment]
Here is my new proposal for this patch:
```diff
From feb3c3b5b65f825de4f9b55175693638c7793587 Mon Sep 17 00:00:00 2001
From: Cala <calaftp@free.fr>
Date: Tue, 5 May 2015 10:37:47 +0200
Subject: [PATCH] Small improvement in Sunken Temple

Shade of Eranikus is unattackable until Jammal'An the Prophet is
defeated
---
 .../sunken_temple/instance_sunken_temple.cpp              | 15 ++++++++++++++-
 1 file changed, 14 insertions(+), 1 deletion(-)

diff --git a/scripts/eastern_kingdoms/sunken_temple/instance_sunken_temple.cpp b/scripts/eastern_kingdoms/sunken_temple/instance_sunken_temple.cpp
index 3e9bfd5..08c31ea 100644
--- a/scripts/eastern_kingdoms/sunken_temple/instance_sunken_temple.cpp
+++ b/scripts/eastern_kingdoms/sunken_temple/instance_sunken_temple.cpp
@@ -87,6 +87,7 @@ void instance_sunken_temple::OnCreatureCreate(Creature* pCreature)
             break;
         case NPC_JAMMALAN:
         case NPC_ATALARION:
+        case NPC_SHADE_OF_ERANIKUS:
             m_mNpcEntryGuidStore[pCreature->GetEntry()] = pCreature->GetObjectGuid();
             break;
     }
@@ -96,13 +97,19 @@ void instance_sunken_temple::OnCreatureEvade(Creature* pCreature)
 {
     switch (pCreature->GetEntry())
     {
-            // Hakkar Event Mobs: On Wipe set as failed!
+        // Hakkar Event Mobs: On Wipe set as failed!
         case NPC_BLOODKEEPER:
         case NPC_HAKKARI_MINION:
         case NPC_SUPPRESSOR:
         case NPC_AVATAR_OF_HAKKAR:
             SetData(TYPE_AVATAR, FAIL);
             break;
+        // Shade of Eranikus: prevent it to become unattackable after a wipe 
+        case NPC_SHADE_OF_ERANIKUS:
+            Creature* pEranikus = GetSingleCreatureFromStorage(NPC_SHADE_OF_ERANIKUS);
+            if (pEranikus)
+                pEranikus->RemoveFlag(UNIT_FIELD_FLAGS, UNIT_FLAG_OOC_NOT_ATTACKABLE);
+            break;
     }
 }
 
@@ -153,6 +160,12 @@ void instance_sunken_temple::SetData(uint32 uiType, uint32 uiData)
             }
             break;
         case TYPE_JAMMALAN:
+            if (uiData == DONE)
+            {
+                Creature* pEranikus = GetSingleCreatureFromStorage(NPC_SHADE_OF_ERANIKUS);
+                if (pEranikus)
+                    pEranikus->RemoveFlag(UNIT_FIELD_FLAGS, UNIT_FLAG_OOC_NOT_ATTACKABLE);
+            }
             m_auiEncounter[uiType] = uiData;
             break;
         case TYPE_AVATAR:
-- 
1.9.5.github.0
```
The results for my tests with this patch after updating UDB with:
```SQL
-- Updated NPC 5709 (Shade of Eranikus) in Sunken Temple: it is now unattackable
-- until Jammal'An is defeated
UPDATE `creature_template` SET `UnitFlags` = `UnitFlags` | 256 WHERE `Entry` = 5709;
```
* If Jammal'An is alive: Eranikus is asleep and does not aggro players even when then come to contact. Players cannot attack Eranikus by any mean
* If Jammal'An is dead:  Eranikus is still asleep and does not aggro players even when then come to contact. Players can engage Eranikus, who fights back
* If players wipe (noobs....): Eranikus goes back to sleep and does not aggro resurected players. Players can engage Eranikus again (and possibly defeat him this time)
